source/adapters/hip/command_buffer.cpp

```diff
@@ -9,156 +9,756 @@
+  // Map of ur_exp_command_buffer_sync_point_t to ur_event_handle_t defining
+  // the event associated with each sync-point
+  auto SyncPoints = CommandBuffer->SyncPoints;
+
+  // For each sync-point add associated HIP graph node to the return list.
+  for (size_t i = 0; i < NumSyncPointsInWaitList; i++) {
+    if (auto NodeHandle = SyncPoints.find(SyncPointWaitList[i]);
+        NodeHandle != SyncPoints.end()) {
+      HIPNodesList.push_back(*NodeHandle->second.get());
+    } else {
+      return UR_RESULT_ERROR_INVALID_VALUE;
+    }
+  }
+  return UR_RESULT_SUCCESS;
+}
+
+// Helper function for enqueuing memory fills
+static ur_result_t enqueueCommandBufferFillHelper(
+    ur_exp_command_buffer_handle_t CommandBuffer, void *DstDevice,
+    const hipMemoryType DstType, const void *Pattern, size_t PatternSize,
+    size_t Size, uint32_t NumSyncPointsInWaitList,
+    const ur_exp_command_buffer_sync_point_t *SyncPointWaitList,
+    ur_exp_command_buffer_sync_point_t *SyncPoint) {
+  ur_result_t Result = UR_RESULT_SUCCESS;
+  std::vector<hipGraphNode_t> DepsList;
+  UR_CALL(getNodesFromSyncPoints(CommandBuffer, NumSyncPointsInWaitList,
+                                 SyncPointWaitList, DepsList),
+          Result);
+
+  try {
+    const size_t N = Size / PatternSize;
+    auto Value = *static_cast<const uint32_t *>(Pattern);
+    auto DstPtr = DstType == hipMemoryTypeDevice
+                      ? *static_cast<hipDeviceptr_t *>(DstDevice)
+                      : DstDevice;
+
+    if ((PatternSize == 1) || (PatternSize == 2) || (PatternSize == 4)) {
+      // Create a new node
+      hipGraphNode_t GraphNode;
+      hipMemsetParams NodeParams = {};
+      NodeParams.dst = DstPtr;
+      NodeParams.elementSize = PatternSize;
+      NodeParams.height = N;
+      NodeParams.pitch = PatternSize;
+      NodeParams.value = Value;
+      NodeParams.width = 1;
+
+      UR_CHECK_ERROR(hipGraphAddMemsetNode(&GraphNode, CommandBuffer->HIPGraph,
+                                           DepsList.data(), DepsList.size(),
+                                           &NodeParams));
+
+      // Get sync point and register the cuNode with it.
+      *SyncPoint = CommandBuffer->AddSyncPoint(
+          std::make_shared<hipGraphNode_t>(GraphNode));
+
+    } else {
+      // HIP has no memset functions that allow setting values more than 4
+      // bytes. UR API lets you pass an arbitrary "pattern" to the buffer
+      // fill, which can be more than 4 bytes. We must break up the pattern
+      // into 1 byte values, and set the buffer using multiple strided calls.
+      // This means that one hipGraphAddMemsetNode call is made for every 1
+      // bytes in the pattern.
+
+      // List to handle inter-node dependencies
+      std::vector<hipGraphNode_t> HIPNodesList = {};
+      // List shared pointer that will point to the last node created
+      std::shared_ptr<hipGraphNode_t> GraphNodePtr;
+
+      size_t NumberOfSteps = PatternSize / sizeof(uint8_t);
```

Comment:
If want 4 bytes, should we be not dividing by size of `uint32_t`?

Reply:
A step consists in adding a node that performs a `uint8_t` memset. 
So, the `NumberOfSteps` must be computed using `uint8_t`.
However, I agree that some fields of `NodeParamsStepFirst` were not set correctly. I updated them.